lib/error.js

```diff
@@ -0,0 +1,63 @@
+/**
+ * @param {string} message error message
+ * @param {"Syntax" | "Validation"} type error type
+ */
+function error(source, position, current, message, type) {
+  const maxTokens = 5; // arbitrary but works well enough
+  const line =
+    source[position].type !== "eof" ? source[position].line :
+    source.length > 1 ? source[position - 1].line :
+    1;
+
+  const precedingLine = lastLine(
+    tokensToText(sliceTokens(-maxTokens), { precedes: true })
+  );
+
+  const procedingTokens = sliceTokens(maxTokens);
+  const procedingText = tokensToText(procedingTokens);
+  const procedingLine = procedingText.split("\n")[0];
+
+  const spaced = " ".repeat(precedingLine.length) + "^ " + message;
+  const contextualMessage = precedingLine + procedingLine + "\n" + spaced;
+
+  const contextType = type === "Syntax" ? "since" : "inside";
+  const grammaticalContext = current ? `, ${contextType} \`${current.partial ? "partial " : ""}${current.type} ${current.name}\`` : "";
+  return `${type} error at line ${line}${grammaticalContext}:\n${contextualMessage}`;
+
+  function sliceTokens(count) {
```

Comment:
style nit: IMO, looks better if these function are hoisted above the actual code and then the last line in the overall function is the `return ...`.

Reply:
I do prefer functions being underneath to allow function bodies to start with actual code, but it seems `parseByTokens()` in lib/webidl2 does return in the last line. I think I should follow the consistency way 😁